### ISSUE TITLE
[BUG] v0.12.1 - The wallet address opens by URL, despite setting a username and password

### ISSUE THREAD

**Issue by Perlover** [lnbits/lnbits#2265]
**Describe the bug**

I did the update. Then I went into the wallet that existed in version 0.11.3 and set a name and password. After that I can login to the wallet with the name and password. Everything is fine here.

But! If I know the URL of the wallet, which was before version 0.12.*, then I can enter it without login! Earlier, when I tested the version on legend.lnbits.com, it worked correctly there, from my point of view - if I entered by URL, I was sent to the login page, because even though I knew the wallet address, it was already protected by name and password.

**To Reproduce**

1. We have an account in version 0.11.3
2. Upgrade to version 0.12.1 (by git's tag)
3. Go to the wallet, set up a name and password for the account. Then logout.
4. For the purity of the experiment, we delete the data and site cookies for the lnbits site.
5. We take the wallet URL address (with usr and wal parameters) and paste it into the browser. The wallet opens immediately, despite the fact that the name and password for it are set.
6. Tested in fresh Firefox version

**Expected behavior**

Earlier, in the dev branch for pre-release 0.12.0, when I was testing - setting a username and password saved from the fact that if someone found out the wallet address - they couldn't log in, because the account was protected by username and password. That's how it should work, in my opinion - and why implement it?

**Desktop (please complete the following information):**

- LNbits version: 0.12.1 (tag "0.12.1")
- Database: sqlite
- Aftet iupdate from v0.11.3

**Comment by Perlover** [comment]
After further investigation of the problem, I realized that the wallet can be opened via a link only when there is a `usr` parameter (we are talking about a user who has protected his account with username and password), or we logged in with username and password (apparently, this method is Access Token).

I believe this is the wrong algorithm for account protection. For example, in versions of LNBits before 0.12.0, wallets always had both parameters in their URL: `usr` & `wal`. If the purpose of adding username and password was to protect old type wallets as well, it failed, because all links of wallets of version < 0.12.0 are opened without password, as they have `usr` in their URLs. You can even consider them permanently compromised. And it is inconvenient to recreate wallets just to create new links without usr parameter.

**My suggestion** is that the LNBits algorithm, having received `usr` from a URL, should check whether the username and password are set for this user or not. If it is, it should request an Access Token (i.e. request a name and password through the login page). If name and password are not specified - open a wallet (to support old accounts where name and password were not set).

**Comment by dadofsambonzuki** [comment]
Currently, you have to disable `user-id-only` in the server config to prevent this.

**Comment by Perlover** [comment]
Thank you! I didn't know about that option. I saw it in the settings.

But doesn't it mean that if I disable `user-id-only`, I won't be able to get into wallets that have `usr` in the URL but no username and password defined?

**Comment by Perlover** [comment]
I have now thoroughly tested with and without the option `user-id-only`.

If I disable the `user-id-only` option, I can't even log into wallets that don't have a username and password set, but have `usr` in the URL. That is, after disabling this option, users who did not have time to set a name and password - lose access to wallets completely!

If I enable the option, it works as I described in this issue.

I would still like to realize the below described option when `user-id-only` and `username-password` options are both enabled:

> **My suggestion** is that the LNBits algorithm, having received usr from a URL, should check whether the username and password are set for this user or not. If it is, it should request an Access Token (i.e. request a name and password through the login page). If name and password are not specified - open a wallet (to support old accounts where name and password were not set).

**Comment by dadofsambonzuki** [comment]
Yes, or at least force the migration process by not allowing access without
creating a username/password.

**Comment by Dominion5254** [comment]
> I have now thoroughly tested with and without the option `user-id-only`.
> 
> If I disable the `user-id-only` option, I can't even log into wallets that don't have a username and password set, but have `usr` in the URL. That is, after disabling this option, users who did not have time to set a name and password - lose access to wallets completely!
> 

Can you share what your `.env` file looked like to produce this result?

I am still able to access with the URL including `usr` even with `user-id-only` removed from `AUTH_ALLOWED_METHODS`:

```
# Auth Configurations
AUTH_SECRET_KEY=""
AUTH_TOKEN_EXPIRE_MINUTES=525600

AUTH_ALLOWED_METHODS="username-password"
```

**Comment by Perlover** [comment]
> Can you share what your `.env` file looked like to produce this result?
> 
> I am still able to access with the URL including `usr` even with `user-id-only` removed from `AUTH_ALLOWED_METHODS`:
> 
> ```
> # Auth Configurations
> AUTH_SECRET_KEY=""
> AUTH_TOKEN_EXPIRE_MINUTES=525600
> 
> AUTH_ALLOWED_METHODS="username-password"
> ```

I have been using the `LNBITS_ADMIN_UI=true` method for a long time, meaning the `.env` file does not reflect the actual settings (settings in sqlite base).

I have the .env file before switching to ADMIN_UI, but the variables you listed are not there. Probably they came later and now their default values are unknown to me.

**Comment by Dominion5254** [comment]
Ah this is very helpful. Thanks.

Following your example above where a username/password are not set, it seems that if the authorization method is set to `username-password` AND a username/password was not set for the Admin account, the Admin could potentially get locked out of their account.

In that situation I believe the only option to regain access would be to manually add `user-id-only` to `auth_allowed_methods` in the `settings` table, then access the Admin account with the `usr` param.

If I am understanding this feature correctly, locking out the admin user (and possibly all other users in the process) seems like a fairly easy footgun to walk into.

**Comment by dni** [comment]
yeah, you shouldn't disable `user-id-only`  on already existing instances with many users, it was meant to be for new isntances else there is no clear migration path for older users.

i see the footgun but i have no idea right now, how we would solve that properly.

@Perlover would adding `AUTH_SECRET_KEY` to admin ui fix your issue?

**Comment by Perlover** [comment]
> i see the footgun but i have no idea right now, how we would solve that properly.

I see the solution described here:

> My suggestion is that the LNBits algorithm, having received usr from a URL, should check whether the username and password are set for this user or not. If it is, it should request an Access Token (i.e. request a name and password through the login page). If name and password are not specified - open a wallet (to support old accounts where name and password were not set).

> 
> @Perlover would adding `AUTH_SECRET_KEY` to admin ui fix your issue?

Honestly, I don't understand the question. I don't know what `AUTH_SECRET_KEY` does. There is no such thing in the UI admin, only in `.env` config but I didn't use it. What does this option do? I am, to be honest, confused in the discussion. I have a feeling that each of us is talking about something else.

**Comment by Perlover** [comment]
Let me try again to express my understanding of the problem.

After upgrading from version 0.11 to 0.12.1, LNBits on the one hand gives users (who already had a wallet) to protect their wallets with a username and password, so that it is impossible to get into old wallet URLs, which theoretically could be compromised (at least that's how I see the purpose of this change to myself, or if the purpose was different - why not follow that purpose as well?).

As a rule, such URLs contained the `usr` parameter. But, after updating and setting a username and password, users who previously had wallets are still vulnerable because their old URLs with the `usr` parameter still open their wallet, regardless of the fact that they have set a username and password.

I suggested above (_"My suggestion is that..."_) to change the algorithm so that previously created wallets can work before the update.

**Comment by motorina0** [comment]
@Perlover valid feature request (not bug), but not a high prio. Will keep the issue open.

**Comment by tlindi** [comment]
This still valid with v0.12.12 - due it is low-prio? (high prio bugs on v0.12.12 said to be maybe fixed?)

**Comment by Perlover** [comment]
Hello everyone! I was the one who opened this ticket, and I've decided to close it because version 1.0 of LNBits has already been released for a while. I think it's probably not relevant to discuss bugs of a version that is no longer supported. Thank you all for your understanding.
